pkg/apis/pipeline/v1alpha1/storage_resource.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2018 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+type PipelineResourceStorageType string
+
+const (
+	// PipelineResourceTypeGCS indicates that resource source is a GCS blob/directory.
+	PipelineResourceTypeGCS PipelineResourceType = "gcs"
+)
+
+// PipelineResourceInterface interface to be implemented by different PipelineResource types
+type PipelineStorageResourceInterface interface {
+	PipelineResourceInterface
+	GetSecretParams() []SecretParam
+	GetDownloadContainerSpec() ([]corev1.Container, error)
+	GetUploadContainerSpec() ([]corev1.Container, error)
+	SetDestinationDirectory(string)
+}
+
+func NewStorageResource(r *PipelineResource) (PipelineStorageResourceInterface, error) {
+	if r.Spec.Type != PipelineResourceTypeStorage {
+		return nil, fmt.Errorf("StoreResource: Cannot create a storage resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "type"):
```

Comment:
in the two scenarios I know about (gs and s3) the type can be derived from the protocol segment of the "location". Do we need type in those cases?

Reply:
`type` here refers to "gcs" and not the type of object downloading. Clients are configured to deduce the format of object from URL

Reply:
the prefix of the location s3://my_bucket/ or gs://my_bucket isn't actually the object type but the protocol specifier which is usually specific to the underlying API. I suppose there may be some storage services that doesn't have a protocol as part of location (azure?) so it may be relevant in such cases. 
In any case, I don't think it's a big deal.